### PR TITLE
Create Collection.group(by:) Method

### PR DESCRIPTION
**Pull request by calebkleveter**
## Checklist

- [x] I checked the [**Contributing Guidelines**](https://github.com/SwifterSwift/SwifterSwift/blob/master/CONTRIBUTING.md) before creating this request.
- [x] New extensions are written in Swift 4.
- [x] New extensions support iOS 8.0+ / tvOS 9.0+ / macOS 10.10+ / watchOS 2.0+.
- [x] I have added tests for new extensions, and they passed.
- [x] All extensions have a **clear** comments explaining their functionality, all parameters and return type in English.
- [x] All extensions are declared as **public**.
- [x] I have added a changelog entry describing my changes.

### CODE REVIEW

File: Sources/Extensions/SwiftStdlib/CollectionExtensions.swift [hunk 1]
```diff
@@ -35,6 +35,18 @@ public extension Collection {
 		}
 	}
 	
+    /// SwifterSwift: Group the elements of the collection in a dictionary.
+    ///
+    ///     [Pair(1, "Hello"), ...].group(by: \Pair.name) -> ["Hello": [Pair(1, "Hello"), ...], ...]
+    ///
+    /// - Parameter key: The `KeyPath` for the element's property to group the collection by.
+    /// - Returns: A dictionary with values grouped with keys.
+    public func group<Value>(by key: KeyPath<Self.Element, Value>) -> [Value: [Self.Element]] {
```
Comment (SD10) [inline Sources/Extensions/SwiftStdlib/CollectionExtensions.swift]: This seems pretty inconsistent with the existing extension on Array. Maybe we can try and formalize these two? Not sure if the Array extension can be moved to Collection.

File: SwifterSwift.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (SD10) [inline SwifterSwift.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist]: Guess this is from Xcode 9.3?

File: Tests/SwiftStdlibTests/CollectionExtensionsTests.swift [hunk 3]
```diff
@@ -31,3 +41,13 @@ final class CollectionExtensionsTests: XCTestCase {
         XCTAssertNil([].randomItem)
     }
 }
+
+struct Pair: Equatable {
```
Comment (SD10) [inline Tests/SwiftStdlibTests/CollectionExtensionsTests.swift]: Can you nest this type in the `testGroupBy()` method please?

File: SwifterSwift.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist (hunk 2 above)
Comment (calebkleveter) [reply SwifterSwift.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist]: Yes, it is.

**Comment by LucianoPAlmeida** [comment]
@SD10 Sure :))

File: Sources/Extensions/SwiftStdlib/ArrayExtensions.swift [hunk 4]
```diff
@@ -358,12 +358,7 @@ public extension Array {
 	/// - Parameter getKey: Clousure to define the key for each element.
 	/// - Returns: A dictionary with values grouped with keys.
 	public func groupByKey<K: Hashable>(keyForValue: (_ element: Element) throws -> K) rethrows -> [K: [Element]] {
-		var group = [K: [Element]]()
-		for value in self {
-			let key = try keyForValue(value)
-			group[key] = (group[key] ?? []) + [value]
-		}
-		return group
+		return try Dictionary(grouping: self, by: keyForValue)
```
Comment (LucianoPAlmeida) [inline Sources/Extensions/SwiftStdlib/ArrayExtensions.swift]: Nice :))

**Comment by omaralbeik** [comment]
Any update on this guys @LucianoPAlmeida, @SD10 :)

**Comment by SD10** [comment]
Honestly, I think that we should just remove the extension `groupByKey` on `Array` and not add this one. This function pretty much just wraps the `Dictionary` initializer. 

The one thing I always hated about `groupByKey` is it gave no context to the return type. What kind of grouping are we doing? A multidimensional array, array of tuples, dictionary? 

I think this is why the `Dictionary` initializer is implemented the way it is.

**Comment by LucianoPAlmeida** [comment]
@SD10 I totally agree, I remember this extension was implemented on back on Swift 3 where the Dictionary initializer did not exist yet, we should totally deprecate it :))

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by calebkleveter** [comment]
Sure @SD10, no problem!

**Comment by calebkleveter** [comment]
Just a thought, should I actually remove `groupByKey`, or just add a `@availible` annotation to it so mark is as deprecated?

**Comment by SD10** [comment]
@calebkleveter You could mark it as deprecated as a better solution. I think there is a `Deprecated` directory in the Swift standard library folder for this. Thanks for clarifying

**Pull request closed without merging**